django/template/library.py

```diff
@@ -0,0 +1,330 @@
+import functools
+import warnings
+from importlib import import_module
+from inspect import getargspec
+
+from django.utils import six
+from django.utils.deprecation import RemovedInDjango21Warning
+from django.utils.itercompat import is_iterable
+
+from .base import Node, Template, token_kwargs
+from .exceptions import TemplateSyntaxError
+
+
+class InvalidTemplateLibrary(Exception):
+    pass
+
+
+class Library(object):
+    """
+    A class for registering template tags and filters. Compiled filter and
+    template tag functions are stored in the filters and tags attributes.
+    The filter, simple_tag, and inclusion_tag methods provide a convenient
+    way to register callables as tags.
+    """
+    def __init__(self):
+        self.filters = {}
+        self.tags = {}
+
+    def tag(self, name=None, compile_function=None):
+        if name is None and compile_function is None:
+            # @register.tag()
+            return self.tag_function
+        elif name is not None and compile_function is None:
+            if callable(name):
+                # @register.tag
+                return self.tag_function(name)
+            else:
+                # @register.tag('somename') or @register.tag(name='somename')
+                def dec(func):
+                    return self.tag(name, func)
+                return dec
+        elif name is not None and compile_function is not None:
+            # register.tag('somename', somefunc)
+            self.tags[name] = compile_function
+            return compile_function
+        else:
+            raise InvalidTemplateLibrary(
+                "Unsupported arguments to Library.tag: (%r, %r)",
```

Comment:
, -> % (missing tests for this branch)
same issue with InvalidTemplateLibrary raised below